### PR TITLE
User namespace networking

### PR DESCRIPTION
**Pull request by rajivnathan**
Adds network policies that allow connections between all namespaces belonging to a user but not from other user's namespaces.

- Adds a new `name` label for each user namespace that will be used for the network selector in the NetworkPolicy
- Adds a NetworkPolicy for each of the user namespaces to allow connections from each of the other namespaces belonging to the same user

Note: I manually verified connectivity between each of the namespace types and also verified other namespaces cannot access the user's namespace.

e2e tests: https://github.com/codeready-toolchain/toolchain-e2e/pull/222

Related issue: https://issues.redhat.com/browse/CRT-819

### CODE REVIEW

File: deploy/templates/nstemplatetiers/advanced/ns_code.yaml [hunk 1]
```diff
@@ -122,6 +124,34 @@ objects:
...
+  spec:
+    ingress:
+      - from:
+        - namespaceSelector:
+            matchLabels:
+              name: ${USERNAME}-dev
+    podSelector: {}
+    policyTypes:
+      - Ingress
+- apiVersion: networking.k8s.io/v1
```
Comment (alexeykazakov) [inline deploy/templates/nstemplatetiers/advanced/ns_code.yaml:141]: Does it have to be a separate policy (one policy per namespace)? Or it's possible to define one policy which would accept traffic from namespaces with either one label **or** another? Basically is it possible to have some kind of **or** operation in the `- from`?

File: pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go [hunk 2]
```diff
@@ -410,10 +410,22 @@ func assertNamespaceTemplate(t *testing.T, decoder runtime.Decoder, actual templ
 	require.NoError(t, err)
 	assert.Equal(t, expected, actual)
 	// Assert expected objects in the template
-	// Each template should have one Namespace, one RoleBinding, one LimitRange and three NetworkPolicy objects
+	// Each template should have one Namespace, one RoleBinding, one LimitRange and a varying number of NetworkPolicy objects depending on the namespace kind
+
+	// Template objects count
+	if tier == "team" {
```
Comment (alexeykazakov) [inline pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go:416]: Let's use switch/case instead.

**Review by alexeykazakov — COMMENTED** [review]
Overall looks good but I have a question.

File: pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go (hunk 2 above)
Comment (rajivnathan) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator_whitebox_test.go]: Done in 95ca667b15c6c242c200cfb797b9eecb577d188c

File: deploy/templates/nstemplatetiers/advanced/ns_code.yaml (hunk 1 above)
Comment (rajivnathan) [reply deploy/templates/nstemplatetiers/advanced/ns_code.yaml:141]: I don't see way to do that. https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#networkpolicyingressrule-v1-networking-k8s-io

Even the label selector only supports ANDing things together: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#labelselector-v1-meta

Comment (alexeykazakov) [reply deploy/templates/nstemplatetiers/advanced/ns_code.yaml:141]: Hm..
> from | List of sources which should be able to access the pods selected for this rule. Items in this list are combined using a logical OR operation. If this field is empty or missing, this rule matches all sources (traffic not restricted by source). If this field is present and contains at least one item, this rule allows traffic only if the traffic matches at least one item in the from list.

Comment (alexeykazakov) [reply deploy/templates/nstemplatetiers/advanced/ns_code.yaml:141]: So sounds like this is what we need:
```
  spec:
    ingress:
      - from:
        - namespaceSelector:
            matchLabels:
              name: ${USERNAME}-dev
        - namespaceSelector:
            matchLabels:
              name: ${USERNAME}-stage
```
right?

Comment (rajivnathan) [reply deploy/templates/nstemplatetiers/advanced/ns_code.yaml:141]: Oops! Yes you're right. I misread it. I'll make the change.

Comment (rajivnathan) [reply deploy/templates/nstemplatetiers/advanced/ns_code.yaml]: Done in https://github.com/codeready-toolchain/host-operator/pull/358/commits/34599fc68836ec12a22b9887db21bf808a7e0423

**Review by alexeykazakov — APPROVED** [review]
Nice!

**Review by MatousJobanek — APPROVED** [review]
Good job :+1:

**Pull request merged**